extensions/mssql/src/credentialstore/nativeCredentialService.ts

```diff
@@ -0,0 +1,144 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { AppContext } from '../appContext';
+import { SqlOpsDataClient, ISqlOpsFeature, SqlOpsFeature } from 'dataprotocol-client';
+import * as Utils from '../utils';
+import { ClientCapabilities, RPCMessageType, ServerCapabilities } from 'vscode-languageclient';
+import * as Contracts from './contracts';
+import { Disposable, SecretStorage } from 'vscode';
+import * as azdata from 'azdata';
+import * as UUID from 'vscode-languageclient/lib/utils/uuid';
+
+export class NativeCredentialService extends SqlOpsFeature<any> {
+
+	private static readonly messagesTypes: RPCMessageType[] = [
+		Contracts.DeleteCredentialRequest.type,
+		Contracts.SaveCredentialRequest.type,
+		Contracts.ReadCredentialRequest.type
+	];
+
+	public static asFeature(context: AppContext): ISqlOpsFeature {
+		return class extends NativeCredentialService {
+			private _secretStorage: SecretStorage;
+			private _useNativeCredentialService: boolean;
+			private _passwordsMigrated: boolean;
+
+			constructor(client: SqlOpsDataClient) {
+				super(context, client);
+				this._secretStorage = context.extensionContext.secrets;
+				this._passwordsMigrated = context.extensionContext.globalState.get(Utils.configPasswordsMigrated);
+				this._useNativeCredentialService = Utils.useNativeCredentialsEnabled();
+			}
+
+			override fillClientCapabilities(capabilities: ClientCapabilities): void {
+				Utils.ensure(Utils.ensure(capabilities, 'credentials')!, 'credentials')!.dynamicRegistration = true;
+			}
+
+			override async initialize(capabilities: ServerCapabilities): Promise<void> {
+				this.register(this.messages, {
+					id: UUID.generateUuid(),
+					registerOptions: undefined
+				});
+				if (this._useNativeCredentialService && !this._passwordsMigrated) {
+					await this.migratePasswords();
+					this._passwordsMigrated = true;
+					await context.extensionContext.globalState.update(Utils.configPasswordsMigrated, this._passwordsMigrated);
+				}
+			}
+
+			/**
+			 * Removes a credential for a given connection profile from sqltoolsservice
+			 * and adds it to secret storage
+			 */
+			private async cleanCredential(conn: azdata.connection.ConnectionProfile): Promise<boolean> {
```

Comment:
`migrateCredential` seems like a better name for this - cleaning it up is only one part of what this function is doing